client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,317 @@
+}
+
+interface TabLink extends HTMLAnchorElement {
+  index: number;
+}
+
+enum Keys {
+  Left = 'ArrowLeft',
+  Right = 'ArrowRight',
+  Home = 'Home',
+  End = 'End',
+}
+
+enum KeyWeight {
+  ArrowLeft = -1,
+  ArrowRight = 1,
+}
+
+/**
+ * @example - creating a simple tabs interface
+ * <div class="w-tabs" data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-active-value="" data-w-tabs-animate-class="animate-in">
+ *   <div class="w-tabs__list" role="tablist" data-w-tabs-target="list">
+ *       <a id="tab-label-tab-1" href="#tab-tab-1" class="w-tabs__tab " role="tab"
+ *          tabindex="-1" data-action="click->w-tabs#handleTabChange:prevent keydown->w-tabs#handleKeydown" data-w-tabs-target="label">
+ *         Tab 1
+ *       </a>
+ *       <a id="tab-label-tab-2" href="#tab-tab-2" class="w-tabs__tab " role="tab"
+ *          data-action="click->w-tabs#handleTabChange:prevent keydown->w-tabs#handleKeydown" data-w-tabs-target="label">
+ *         Tab 2
+ *       </a>
+ *   </div>
+ *
+ *   <div class="tab-content tab-content--comments-enabled">
+ *       <section id="tab-tab-1" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
+ *           tab-1
+ *       </section>
+ *       <section id="tab-tab-2" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
+ *           tab-2
+ *       </section>
+ *   </div>
+ * </div>
+ */
+
+export class TabsController extends Controller<HTMLDivElement> {
+  static targets = ['list', 'trigger', 'label', 'panel'];
+
+  static classes = ['animate'];
+
+  static values = {
+    transition: { default: 150, type: Number },
+    active: { default: '', type: String },
+    disable: { default: false, type: Boolean },
+    animate: { default: true, type: Boolean },
+  };
+
+  declare animateClass: string;
+
+  declare listTarget: HTMLDivElement;
+  declare triggerTargets: HTMLAnchorElement[];
+  declare labelTargets: HTMLAnchorElement[];
+  declare panelTargets: HTMLElement[];
+
+  declare disableValue: boolean;
+  declare activeValue: string;
+  declare transitionValue: number;
+  declare animateValue: boolean;
+
+  activeValueChanged(currentValue: string, previousValue: string) {
+    if (previousValue) {
+      this.hideTabContent(previousValue);
+    }
+
+    const tab = this.getTabLabelByHref(currentValue);
+    if (tab) {
+      tab.setAttribute('aria-selected', 'true');
+      tab.removeAttribute('tabindex');
+    }
+
+    const tabContent = this.getTabPanelByHref(currentValue);
+
+    if (tabContent) {
+      if (this.animateValue) {
+        this.animateIn(tabContent);
+      } else {
+        tabContent.hidden = false;
+      }
+
+      this.listTarget.dispatchEvent(
+        new CustomEvent('switch', {
+          detail: { tab: tab?.getAttribute('href')?.replace('#', '') },
+        }),
+      );
+      document.dispatchEvent(new CustomEvent('wagtail:tab-changed'));
```

Comment:
When dispatching events from Stimulus, you can use `this.dispatch`. 

```suggestion
      this.dispatch('changed', { cancelable: false, detail: { to: previouValue }});
```

This will dispatch an event that bubbles with the name `w-tabs:changed`. 

Any other code in Wagtail that relies on the current event name will need to be updated. We will also add upgrade considerations for this when it's closer to being merged. 

See https://stimulus.hotwired.dev/reference/controllers#cross-controller-coordination-with-events